07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+jlToList (dropJ n jl) == drop n (jlToList jl)
+-}
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ n (Single m a)
+    | n >= 1    = Empty
+    | otherwise = Single m a
+dropJ n (Append _ jl1 jl2)
+    | n < jl1Size = dropJ n             jl1 +++ jl2
+    | otherwise   = dropJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+dropJ _ _ = Empty
+
+
+
+
+{-
+Finally, implement the function takeJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The takeJ function returns the first n elements of a JoinList, dropping all other elements. Again, this function works similarly to the standard library take function; that is, it should be the case that
+jlToList (takeJ n jl) == take n (jlToList jl).
+Ensure that your function definitions use the size function from the Sized type class to make smart decisions about how to descend into the JoinList tree.
+-}
+takeJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+takeJ n (Single m a)
+    | n <= 0    = Empty
+    | otherwise = Single m a
+takeJ n (Append _ jl1 jl2)
+    | n < jl1Size = takeJ n jl1
+    | otherwise   = jl1 +++ takeJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+takeJ _ _ = Empty
+
+--Exercise 3
+scoreLine :: String -> JoinList Score String
+scoreLine str = Single (scoreString str) str
+
+
+{-
+Exercise 4 Finally, combine these two kinds of annotations. A pair of monoids is itself a monoid:
+ instance (Monoid a, Monoid b) => Monoid (a,b) where
+   mempty = (mempty, mempty)
+   mappend (a1,b1) (a2,b2) = (mappend a1 a2, mappend b1 b2)
+(This instance is defined in Data.Monoid.) This means that join-lists can track more than one type of annotation at once, in parallel, sim- ply by using a pair type.
+Since we want to track both the size and score of a buffer, you should provide a Buffer instance for the type
+JoinList (Score, Size) String.
+Due to the use of the Sized type class, this type will continue to work with your functions such as indexJ.
+Finally, make a main function to run the editor interface using your join-list backend in place of the slow String backend (see StringBufEditor.hs for an example of how to do this). You should create an initial buffer of type JoinList (Score, Size) String and pass it as an argument to runEditor editor. Verify that the editor demonstration described in the section “Editors and Buffers” does not exhibit delays when showing the prompt.
+-}
+--Exercise 4
+instance Buffer (JoinList (Score, Size) String) where
+    toString = unlines . fromDList . jlToDList
+
+    fromString = foldl' func Empty . lines
+      where
+        func :: JoinList (Score, Size) String -> String -> JoinList (Score, Size) String
+        func jl str = jl +++ Single (scoreString str, 1) str
```

Comment:
This function could've been implemented using the lambda notation (as is idiomatic for fold functions) since it is short enough. Is there any specific reason why you?

Reply:
@oishikg I think I didn't want the line to be too long which is why I abstracted it out into another function called `func`.